[TF FE] Fix centernet and correct FloorDiv translator for signed integer type

pavel-esir opened the pull request:
### Details:
 - Centernet's topk operation returns large int32 values (greater than 1000000), even though they are integer `FloorDiv`/`Div(inp_1, inp_2) + Floor` operation is performed in float16 and because of that it causes accuracy problems.

 - To solve this need to performs FloorDiv operation in integer with a subgraph:
```
res = x / y; if x > 0 and y > 0
res = x / y - 1; if (x < 0 xor y < 0) and (x mod y != 0)
```
 - checked on separate bus: no degradations caused.

### Tickets:
 - CVS-130526

rkazants commented on src/frontends/tensorflow_common/src/op/binary_op.cpp:
```diff
@@ -57,8 +58,16 @@ OutputVector translate_floor_div_op(const NodeContext& node) {
+            auto divide_node = make_shared<v1::Divide>(float_x, float_y);
+            auto floor_node = make_shared<v0::Floor>(divide_node);
+            auto convert_node = make_shared<v0::Convert>(floor_node, out_type);
+            disable_fp16_compression(float_x);
+            disable_fp16_compression(float_y);
+            disable_fp16_compression(divide_node);
+            disable_fp16_compression(floor_node);
+            disable_fp16_compression(convert_node);
```
these are strange changes a bit. During FE conversion we work with the original type. Here I don't think we need any compression specific logic

pavel-esir replied:
I agree that it looks a bit strange. Reason i tried such solution is that we already had a similar thing for random normal
https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/common/src/random_normal_helper.cpp#L60-L62
But still it's strange, and moreover for general solution we should keep precision for all FWs.

I am currently rewriting this solution, in the latest version i call a separate pass inside `save_model`. I will write when it's ready

pavel-esir replied:
@rkazants rewritten solution. Please review

pavel-esir commented on tests/layer_tests/tensorflow_tests/test_tf_FloorDiv.py:
```diff
@@ -17,9 +16,8 @@ def create_add_placeholder_const_net(self, x_shape, dtype, ir_version, use_new_f
-            constant_value = np.array(-10).astype(dtype)
+            constant_value = np.array(-256).astype(dtype)
             y = tf.constant(constant_value)
-            x = tf.raw_ops.Abs(x=x)
```
this abs is not necessary here, and it causes problems when GPU inference is checked
![image](https://github.com/openvinotoolkit/openvino/assets/5703039/aac283d5-8a43-49eb-ac87-1c9f1944ca3c)

popovaan commented on tests/layer_tests/tensorflow_tests/test_tf_FloorDiv.py:
```diff
@@ -17,9 +16,8 @@ def create_add_placeholder_const_net(self, x_shape, dtype, ir_version, use_new_f
-            constant_value = np.array(-10).astype(dtype)
+            constant_value = np.array(-256).astype(dtype)
             y = tf.constant(constant_value)
-            x = tf.raw_ops.Abs(x=x)
```
Why is this change needed?

rkazants replied:
Agree, let us fix it as in a separate PR.

pavel-esir commented:
@popovaan @rkazants i reverted removing of abs and instead added a separate tests cases where i statically specify inputs instead of random generation. I found that even for a very simple cases the current main, e.g. 
FloorDiv(10, 10) which is equivalent to python 10 // 10 should be 1, but after compression of IR OpenVINO gives answer 0, 

![image](https://github.com/openvinotoolkit/openvino/assets/5703039/9457d168-6a71-422f-8591-cd4461a5b724)

because if this denominator 10 is converted to Div(input_1, 10) => Mul(input_1, 1/10) then when 1/10 = 0.1 is compressed to f16 it becomes 0.099975.., which causes difference even for CPU inference!!! (because floor(10 * 0.099..) = floor(0.99..) = 0)

![image](https://github.com/openvinotoolkit/openvino/assets/5703039/f76c8dc6-4612-4129-80e4-c80ac022257b)
But after this fix works fine, since this constant is stored exactly without compression.

One more thing, since layer test call `mo` instead of `ovc` `save_model` is not called in layer tests, and therefore even with my fix tests appeared red. I added some changes to layer test class so that `ovc` is called, if these changes are radical i will revert them.

itikhono commented on src/common/transformations/include/transformations/rt_info/disable_fp16_compression.hpp:
```diff
@@ -39,7 +39,7 @@ class TRANSFORMATIONS_API DisableFP16Compression : public RuntimeAttribute {
     }
 
     bool is_copyable() const override {
-        return false;
```
looks not safe
As far as I remember, this attribute affects several transformations
Did we run any validation for this change?

pavel-esir replied:
it was needed in my very fist draft solution but now it's not necessary. 

Ok, i got it, it's dangerous to change it now: we have a separate ticket for making in copyable CVS-130863

itikhono commented on src/common/transformations/src/transformations/fp16_compression/mark_subgraphs_to_keep_in_mixed_precision.cpp:
```diff
@@ -423,6 +424,35 @@ class PropagateDownDisableSensitivityForQuantized : public pass::MatcherPass {
+        // Floor(Div(input_1, input_2))
+        // Floor(Mul(input_1, inversed_const_input_2)), if input_2 is constant
+        auto div_mul_pattern = pattern::wrap_type<ov::op::v1::Divide, ov::op::v1::Multiply>();
+        auto floor_pattern = pattern::wrap_type<ov::op::v0::Floor>({div_mul_pattern});
```
should we check that 2nd input to Mul is constant according to the comment?
 // Floor(Mul(input_1, inversed_const_input_2)), if input_2 is constant

pavel-esir replied:
Agreed, i corrected the pattern. No need to check Const values/types, they should be marked in any case.

rkazants commented on tests/layer_tests/common/utils/common_utils.py:
```diff
@@ -48,15 +48,36 @@ def generate_ir(coverage=False, **kwargs):
+    from openvino.tools import mo, ovc
+    
     out_dir = kwargs['output_dir'] + os.sep + kwargs['model_name'] + ".xml"
-
-    # TODO: Remove usage of legacy params from layer tests and switch to convert_model from tools.ovc
-    ov_model = convert_model(**kwargs)
-    serialize(ov_model, out_dir)
-
+    compress_to_fp16 = getattr(kwargs, 'compress_to_fp16', True)
+
+    import inspect
+    acceptable_mo_arg_names = ['use_new_frontend', 'model_name', 'compress_to_fp16', 'output_dir']  # mo args that do not conflict with ovc
+    ovc_arg_names = inspect.signature(ovc.convert_model).parameters.keys()
+
+    # By default, run in ovc, but fallback to mo if there are args not present in ovc.convert_model.
+    # Also, if there are such arguments as 'model_name', 'out_dir', etc. they do not conflict with ovc, they should be just
+    # removed from kwargs and then ovc.convert_model can be run with updated ovc_kwargs.
+    use_ovc = True
+    
+    ovc_kwargs = kwargs.copy()
+    for key in kwargs.keys():
+        if key not in ovc_arg_names:
+            ovc_kwargs.pop(key)
+        if key not in ovc_arg_names and key not in acceptable_mo_arg_names:
+            use_ovc = False
+            break
+
+    if use_ovc:
+        ov_model = ovc.convert_model(**ovc_kwargs)    
+    else:
+        # TODO: Remove usage of legacy params from layer tests and switch completely to ovc.convert_model
+        ov_model = mo.convert_model(**kwargs)    
+    
+    save_model(ov_model, out_dir, compress_to_fp16)
```
Why do we need these changes? Let us do this in separate PR

pavel-esir replied:
Since marking is called inside `save_model` it was needed to call `save_model` in layer tests so that this fix work. I tried to call `ovc + save_model` instead of `mo + serialize` in cases where it's possible. But since it didn't work well, i reverted it.
Agreed, let's consider using ovc in a separate PR.

Now i always call `mo + save_model`. @popovaan, could you please confirm that always calling `save_model` is ok?

pavel-esir replied:
Fully reverted common layer test changes. Switching to ovc will be done under CVS-132151

Xfailed my newly added tests since for them to be green `save_model` should be called which impossible to call without common layer test changes. When i run locally FloorDiv layer tests with `save_model` they are green
![image](https://github.com/openvinotoolkit/openvino/assets/5703039/01e6436b-2e8d-42d7-9ac0-efb339cfcb0f)

rkazants reviewed (changes requested):
let us avoid common layer test changes

pavel-esir commented:
> let us avoid common layer test changes

@rkazants agreed and reverted these changes back.

pavel-esir commented on src/common/transformations/src/transformations/fp16_compression/mark_subgraphs_to_keep_in_mixed_precision.cpp:
```diff
@@ -420,6 +421,41 @@ class PropagateDownDisableSensitivityForQuantized : public pass::MatcherPass {
+        MATCHER_SCOPE(MarkFloorDiv);
+        // FloorDiv is repesented either as:
+        // Floor(Div(input_1, input_2))
+        // Floor(Mul(input_1, inversed_input_2)), if input_2 is constant
```
```suggestion
        // Floor(Mul(input_1, inversed_const_input_2)), if input_2 is a constant
```

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_FloorDiv.py:
```diff
@@ -49,3 +48,65 @@ def test_add_placeholder_const_1D(self, params, ie_device, precision, ir_version
+        import tensorflow as tf
+        
+        x = np.arange(min, max, step, dtype=dtype)
+        x_shape = x.reshape(kwargs['x_shape']).shape if 'x_shape' in kwargs else x.shape
```
please extract shape from input_info dictionary (like it is done in other layers tests) instead of using kwargs.
I am afraid modifying kwarg can affect other test cases

pavel-esir replied:
unfortunately until the IR is not created we don't have `input_info` dictionary id didn't modify here `kwargs`, we are only reading it and it's not modified anywhere else

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_FloorDiv.py:
```diff
@@ -49,3 +48,65 @@ def test_add_placeholder_const_1D(self, params, ie_device, precision, ir_version
+    
+    def _prepare_input(self, inputs_dict, kwargs):
+        for input in inputs_dict.keys():
+            inputs_dict[input] = np.arange(kwargs['min'], kwargs['max'], kwargs['step'], dtype=kwargs['dtype'])
```
please save `min` and other values in self corresponding the object. Avoid modifying `kwargs`

pavel-esir replied:
but i didn't modify `kwargs` at all. i just pass them with the currently existing mechanism with `kwargs_to_prepare_input`: https://github.com/openvinotoolkit/openvino/blob/7423fc52a42a441509727b06c0907143a3481a51/tests/layer_tests/common/layer_test_class.py#L88-L92

It's already used e.g.
https://github.com/openvinotoolkit/openvino/blob/master/tests/layer_tests/tensorflow_tests/test_tf_FakeQuantize.py#L142-L148
https://github.com/openvinotoolkit/openvino/blob/7423fc52a42a441509727b06c0907143a3481a51/tests/layer_tests/pytorch_tests/test_arange.py#L121-L123

pavel-esir replied:
i would've been ideal to do like this without mutable dicts, 
![image](https://github.com/openvinotoolkit/openvino/assets/5703039/5e5074c9-c73d-436f-aa2f-7e658ed68c84)

but unfortunately in order to do so, we should modify common tests here  `kwargs['kwargs_to_prepare_input']` -> `**kwargs['kwargs_to_prepare_input']` https://github.com/openvinotoolkit/openvino/blob/7423fc52a42a441509727b06c0907143a3481a51/tests/layer_tests/common/layer_test_class.py#L88-L90

like it's for torch layer tests e.g. in torch 
https://github.com/openvinotoolkit/openvino/blob/7423fc52a42a441509727b06c0907143a3481a51/tests/layer_tests/pytorch_tests/pytorch_layer_test_class.py#L51

Can we consider this improvements separately and merge this PR as is?

rkazants replied:
We have one day before code freeze and this change should take less one hour. So please fix it in this PR.

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_FloorDiv.py:
```diff
@@ -49,3 +48,65 @@ def test_add_placeholder_const_1D(self, params, ie_device, precision, ir_version
+
+    test_inputs = [
+        # test for integers
+        dict(min=-20, max=20, step=1, y=[10],dtype=np.int32),
+        dict(min=-10000, max=10000, step=100, y=[10000], dtype=np.int32),
+        dict(min=-10000, max=10000, step=100, y=[-10000], dtype=np.int32),
+        dict(min=-1e5, max=1e5, step=100, y=[1e5], dtype=np.int32),
+        dict(min=-1e8, max=1e8, step=1e4, y=[1e5], dtype=np.int32),
+        dict(min=-1e8, max=1e8, step=1e4, y=[-1e5], dtype=np.int32),
+        
+        # test for multidimensinal input
+        dict(min=-10000, max=10000, step=10, y=[10000], x_shape=[20, -1], dtype=np.int32),
+        dict(min=-10000, max=10000, step=10, y=[10000], x_shape=[2, 5, -1], dtype=np.int32),
+        dict(min=-10000, max=10000, step=1, y=[10000], x_shape=[2, 5, 10, -1], dtype=np.int32),
+
+        # test for floats
+        dict(min=-20, max=20, step=1, y=[10],dtype=np.float32),
+        dict(min=-10000, max=10000, step=100, y=[10000], dtype=np.float32),
+        dict(min=-10, max=10, step=10, y=[100], dtype=np.float64),
+        dict(min=-1e5, max=1e5, step=100, y=[1e5], dtype=np.float32),
+        dict(min=-1e8, max=1e8, step=1e4, y=[1e5], dtype=np.float32),
+        dict(min=-1e8, max=1e8, step=1e4, y=[-1e5], dtype=np.float32),
```
Is it reasonable to use such extreme values here: 10000 and -1e5? Are we sure to avoid sporadic issues?

pavel-esir replied:
indeed i had quite a big values around `1e6` on FloorDiv input
![image](https://github.com/openvinotoolkit/openvino/assets/5703039/9119da5f-5938-4f60-8d0e-96a185b2866e)

but cases for `1e8` are probably redundant i will replace them with values  `~100-1000`

pavel-esir commented:
@rkazants implemented with ints. Please review

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_FloorDiv.py:
```diff
@@ -49,3 +48,75 @@ def test_add_placeholder_const_1D(self, params, ie_device, precision, ir_version
+        dict(min=-1000, max=1000, step=2, y=[100], dtype=np.int32),
+        dict(min=-10000, max=10000, step=100, y=[10000], dtype=np.int32),
+        dict(min=-10000, max=10000, step=100, y=[-10000], dtype=np.int32),
+        dict(min=-1e5, max=1e5, step=100, y=[1e5], dtype=np.int32),
```
this case is heavy for layer tests. Let us remove it.

rkazants replied:
and dtype can be parameterized to have int32 and int64

pavel-esir replied:
> this case is heavy for layer tests because of big size. Let us remove it.

it might look heavy, but indeed it's not heavy at all. it's just 1000 elements, and even when i increate it by 1000 times to 1e8 execution time still is around ~4 secs!!! I left it just to be sure about large numbers. But removed other cases

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_FloorDiv.py:
```diff
@@ -49,3 +48,75 @@ def test_add_placeholder_const_1D(self, params, ie_device, precision, ir_version
+    test_inputs = [
+        # test for integers
+        dict(min=-20, max=20, step=1, y=[10],dtype=np.int32),
+        dict(min=-200, max=200, step=1, y=[-100], dtype=np.int32),
+        dict(min=-1000, max=1000, step=2, y=[100], dtype=np.int32),
+        dict(min=-10000, max=10000, step=100, y=[10000], dtype=np.int32),
+        dict(min=-10000, max=10000, step=100, y=[-10000], dtype=np.int32),
+        dict(min=-1e5, max=1e5, step=100, y=[1e5], dtype=np.int32),
+        dict(min=1e8, max=1e9, step=1e8, y=[1e8], dtype=np.int64),
+        dict(min=1e8, max=1e9, step=1e8, y=[-1e8], dtype=np.int64),
+        
+        # test for multidimensinal input
+        dict(min=-1000, max=1000, step=10, y=[1000], x_shape=[20, -1], dtype=np.int32),
+        dict(min=-1000, max=1000, step=10, y=[1000], x_shape=[2, 5, -1], dtype=np.int32),
+        dict(min=-1000, max=1000, step=1, y=[1000], x_shape=[2, 5, 10, -1], dtype=np.int32),
+
+        # test for floats
+        dict(min=-10, max=10, step=1, y=[10], dtype=np.float32),
+        dict(min=-1e5, max=1e5, step=100, y=[1e5], dtype=np.float64),
```
can we remove all test cases expect the first one and add cases when y=[-5], y=[-6], y=[5], y=[6]?

rkazants replied:
floating-point does not makes sense to verify. it is already checked.

rkazants replied:
Now I am afraid we have execution time increase of layer tests due to big size tensors

pavel-esir replied:
> Now I am afraid we have execution time increase of layer tests due to big size tensors

it might look heavy, but indeed it's not heavy at all. it's just 1000 elements, and even when i increate it by 1000 times to 1e8 execution time still is around ~4 secs!!! But anyways i removed other cases. Left only one for 1e5 just to be sure about large numbers.

rkazants commented on src/frontends/tensorflow_common/src/op/binary_op.cpp:
```diff
@@ -55,10 +57,25 @@ OutputVector translate_floor_div_op(const NodeContext& node) {
+            auto div = make_shared<v1::Divide>(x, y, true);
+
+            auto zero_const = make_shared<v0::Constant>(out_type, Shape{}, 0);
+            auto minux_one_const = make_shared<v0::Constant>(out_type, Shape{}, -1);
```
```suggestion
            auto minus_one_const = make_shared<v0::Constant>(out_type, Shape{}, -1);
```

pavel-esir replied:
🤦 sorry. fixed that.

rkazants commented on src/frontends/tensorflow_common/src/op/binary_op.cpp:
```diff
@@ -55,10 +57,25 @@ OutputVector translate_floor_div_op(const NodeContext& node) {
+            auto cond_mod = make_shared<v1::NotEqual>(mod_xy, zero_const);
+
+            auto cond = make_shared<v1::LogicalAnd>(cond_mod, xor_cond);
+            auto reminder = make_shared<v1::Select>(cond, minux_one_const, zero_const);
```
```suggestion
            auto reminder = make_shared<v1::Select>(cond, minus_one_const, zero_const);
```

rkazants commented:
Layer test is failing:

```
FAILED install/tests/layer_tests/tensorflow_tests/test_tf_FloorDiv.py::TestFloorDivStaticInput::test_floordiv[ ie_device:CPU - precision:FP16 - params:{'min': -10000, 'max': 10000, 'step': 100, 'y': [10000], 'dtype': <class 'numpy.int32'>} ] - AssertionError: Comparing with Framework failed: ie_res={'FloorDiv:0': array([-1, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2,
       -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2,
       -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2,
       -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2,
       -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2,
       -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2, -2,  0,  0,
        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0], dtype=int32)}; framework_res={'FloorDiv:0': array([-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
       -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
       -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
       -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
       -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
       -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  0,  0,
        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
        0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0], dtype=int32)}.
```

pavel-esir commented:
> Layer test is failing:

It turns out that ARM performs Div a bit differently. And you already faced it here https://github.com/openvinotoolkit/openvino/pull/21730#issuecomment-1879699294
I opened ticket for that CVS-132377

Assuming from the outputs `Div` operation on ARM is already `FloorDiv` (or as it should been theoretically when `pythondiv=true` but this flag as i see does not work).

rkazants commented:
> > Layer test is failing:
> 
> It turns out that ARM performs Div a bit differently. And you already faced it here [#21730 (comment)](https://github.com/openvinotoolkit/openvino/pull/21730#issuecomment-1879699294) I opened ticket for that [CVS-132377](https://jira.devtools.intel.com/browse/CVS-132377)
> 
> Assuming from the outputs `Div` operation on ARM is already `FloorDiv` (or as it should been theoretically when `pythondiv=true` but this flag as i see does not work).

Can you think of how to bypass this now? For example, make both `x` and `y` to be non-negative then compute divide and adjust this result to negative case. Now `Divide` behaves differently on various platforms for negative remainder. Is it possible?

Best regards,
Roman

pavel-esir commented:
> Can you think of how to bypass this now? For example, make both `x` and `y` to be non-negative then compute divide and adjust this result to negative case. Now `Divide` behaves differently on various platforms for negative remainder. Is it possible?
> 
> Or maybe `FloorMod` will help. Compute `FloorMod` and then compute `Divide`. I don't think [CVS-132377](https://jira.devtools.intel.com/browse/CVS-132377) is totally blocking us, try to find another correct solution.

Unfortunately `FloorMod` didn't help. I tried a bunch of other options: e.g. `(2 * x + 1) / 2` which should works differently on ARM and add reminder depending on this but it didn't help, some other options didn't help as well.

The best option i found is to add subgraph with `Select`, `Ceil`, `Floor`,.. not on FE but selectively only for `CPU`, `GPU` without `ARM` on transformations phase. @rkazants could you please take a look to the solution?

pavel-esir commented:
@alvoron FYI since Div on Arm returns values different from cpu/gpu this FloorDiv also differ. xfailed Div and FloorDiv layer tests for arm until CVS-132377 is resolved.

rkazants commented on src/frontends/tensorflow_common/src/op/binary_op.cpp:
```diff
@@ -54,11 +57,25 @@ OutputVector translate_binary_op(const NodeContext& node,
+            auto cond = make_shared<v1::LogicalAnd>(cond_mod, xor_cond);
+            auto reminder = make_shared<v1::Select>(cond, minus_one_const, zero_const);
+            return make_shared<v1::Add>(div, reminder);
         } else {
```
please add separate branch for unsigned int that uses only `Divide`

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_FloorDiv.py:
```diff
@@ -3,23 +3,25 @@
+rng = np.random.default_rng()
 
 class TestFloorDiv(CommonTFLayerTest):
+    dtype = np.int32
+
```
please remove it

rkazants replied:
```suggestion
```

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_FloorDiv.py:
```diff
@@ -45,7 +63,66 @@ def create_add_placeholder_const_net(self, x_shape, dtype, ir_version, use_new_f
     def test_add_placeholder_const_1D(self, params, ie_device, precision, ir_version, temp_dir,
                                       use_new_frontend):
+        
+        if platform.machine() == 'arm64' and np.issubdtype(params['dtype'], np.integer):
```
```suggestion
        if platform.machine() == 'arm64' and np.issubdtype(params['dtype'], np.signedinteger):
```

pavel-esir replied:
done

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_FloorDiv.py:
```diff
@@ -45,7 +63,66 @@ def create_add_placeholder_const_net(self, x_shape, dtype, ir_version, use_new_f
     @pytest.mark.precommit_tf_fe
     def test_add_placeholder_const_1D(self, params, ie_device, precision, ir_version, temp_dir,
                                       use_new_frontend):
+        
```
```suggestion
```

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_FloorDiv.py:
```diff
@@ -45,7 +61,65 @@ def create_add_placeholder_const_net(self, x_shape, dtype, ir_version, use_new_f
+    @pytest.mark.parametrize("dtype", [np.int32, np.int64])
+    @pytest.mark.nightly
+    @pytest.mark.precommit_tf_fe
+    @pytest.mark.skipif(platform.machine() == 'arm64', reason = 'Ticket CVS-132377 - Divide inconsistent behavior on different systems')
```
```suggestion
    @pytest.mark.xfail(condition=platform.system() == 'Linux' and platform.machine() in ['arm', 'armv7l',
                                                                                         'aarch64',
                                                                                         'arm64', 'ARM64'],
                       reason='Ticket CVS-132377 - Divide inconsistent behavior on different systems')
```

pavel-esir replied:
many thanks, this is exactly what i needed

pavel-esir commented on tests/layer_tests/tensorflow_tests/test_tf_FloorDiv.py:
```diff
@@ -45,7 +61,65 @@ def create_add_placeholder_const_net(self, x_shape, dtype, ir_version, use_new_f
     @pytest.mark.precommit_tf_fe
     def test_add_placeholder_const_1D(self, params, ie_device, precision, ir_version, temp_dir,
                                       use_new_frontend):
+        if platform.machine() == 'arm64' and np.issubdtype(params['dtype'], np.signedinteger):
```
```suggestion
        if platform.system() == 'Linux' and platform.machine() in ['arm', 'armv7l', aarch64', 'arm64', 'ARM64'] and np.issubdtype(params['dtype'], np.signedinteger):
```
